**Michał Sitko, Michał Chmielowiec, Andrzej Dawidziuk**
# Neural Networks (SNR) 2018L - Final Report


---

We decided to use [Google Collaboratory](https://colab.research.google.com) as a main platform for project development. So called Colab was developed by Google internally from year 2014 as a tool dedicated for machine learning reaserch. Now it is publicly available as a side Google Drive service.

Colab is an online version of Jupyter Notebook - a web application that allows its users to create and share documents that contain live code, equations, visualizations and narrative text. Usually Jupyter Notebooks are hosted locally on researchers' machines. Very often these machines are not equipped with processing units suited for complex calculations. Locally prepared Jupyter reports are printed as PDF documents and then statically distributed.

Google Colab provides online version of Jupyter server with exposed by very convenient web interface. Each Google Drive account is allowed to use one Jupyter Notebook environment. Technically, environments are setup inside cloud hosted, UNIX type, Docker containers. The containers' virtual hard drives can expand to the maximum space available for particular account in its Google Drive service. Furthermore the container environment has access to pretty powerful computing resources: quad core CPU and Tesla K80 GPU. We should note here that those resources are shared by multiple users and thus its base performance may vary.

*This report was prepared in Colab.*

## 1. Environment setup

Let's ensure that the installed Python version is 2.7. It happens occasionally that containers are initialized with Python 3.6 kernels despite the fact that they the lower one is configured. In such a situation the hypervisor should be rebooted (i.e. by invoking a kill command).

In [0]:
# If python2 is not loaded kill the hypervisor
# ! kill -9 -1
import sys
sys.version

'2.7.14 (default, Sep 23 2017, 22:06:14) \n[GCC 7.2.0]'

Thanks to Colab platform, GPU device is available!

In [0]:
import tensorflow as tf
tf.test.gpu_device_name()

'/device:GPU:0'

By calling appropriate backend methods we can get some more specific info concerning processing devices that are available.

In [0]:
from tensorflow.python.client import device_lib
device_lib.list_local_devices()

[name: "/device:CPU:0"
 device_type: "CPU"
 memory_limit: 268435456
 locality {
 }
 incarnation: 14074366061964417666, name: "/device:GPU:0"
 device_type: "GPU"
 memory_limit: 11287966516
 locality {
   bus_id: 1
   links {
   }
 }
 incarnation: 13993350310858143274
 physical_device_desc: "device: 0, name: Tesla K80, pci bus id: 0000:00:04.0, compute capability: 3.7"]

Now we will clone the project repository and install required pip packages. Interactive shell can be accessed inline by prefixing commands with an exclamation mark (!).

In [0]:
% cd ~

# Remove the environment
! if [ -d "nn-nbirds" ]; then rm -rf "nn-nbirds"; fi
# ! pip freeze | xargs pip uninstall -y

# Build the environment 
! git clone https://github.com/zacateras/nn-nbirds.git
% cd ./nn-nbirds
! pip install -r requirements.txt > pip.log
# The commands below fix the issue with Keras @ Colab
! pip install Pillow==4.0.0 -q
! pip install PIL -q
! pip install image -q

/content
Cloning into 'nn-nbirds'...
remote: Counting objects: 310, done.
remote: Total 310 (delta 0), reused 0 (delta 0), pack-reused 310
Receiving objects: 100% (310/310), 548.66 MiB | 33.45 MiB/s, done.
Resolving deltas: 100% (134/134), done.
Checking out files: 100% (81/81), done.
/content/nn-nbirds
  Could not find a version that satisfies the requirement ipython==6.3.1 (from -r requirements.txt (line 15)) (from versions: 0.10, 0.10.1, 0.10.2, 0.11, 0.12, 0.12.1, 0.13, 0.13.1, 0.13.2, 1.0.0, 1.1.0, 1.2.0, 1.2.1, 2.0.0, 2.1.0, 2.2.0, 2.3.0, 2.3.1, 2.4.0, 2.4.1, 3.0.0, 3.1.0, 3.2.0, 3.2.1, 3.2.2, 3.2.3, 4.0.0b1, 4.0.0, 4.0.1, 4.0.2, 4.0.3, 4.1.0rc1, 4.1.0rc2, 4.1.0, 4.1.1, 4.1.2, 4.2.0, 4.2.1, 5.0.0b1, 5.0.0b2, 5.0.0b3, 5.0.0b4, 5.0.0rc1, 5.0.0, 5.1.0, 5.2.0, 5.2.1, 5.2.2, 5.3.0, 5.4.0, 5.4.1, 5.5.0, 5.6.0, 5.7.0)
No matching distribution found for ipython==6.3.1 (from -r requirements.txt (line 15))
  Could not find a version that satisfies the requirement PIL (from versions: )
No m

The whole **nbirds** database was stored in dropbox as a zip blob. We prepared automated shell script for downloading and unzipping it. Let's use it now and split the dataset stably into train *(0.7)*, validation *(0.2)* and test parts *(0.1)*. Very high download speed (>30MB/s) and connection stability provided by the service should be noted here.

In [0]:
# Download the dataset
! ./data_tools/download.sh

# Split the dataset
! ./data_tools/split.sh

--2018-06-12 17:59:53--  https://www.dropbox.com/s/fi2g3zxsn0pdmn1/nbirds.zip
Resolving www.dropbox.com (www.dropbox.com)... 162.125.9.1, 2620:100:601f:1::a27d:901
Connecting to www.dropbox.com (www.dropbox.com)|162.125.9.1|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://ucdb7c855b5e2daa601e6fc04f1d.dl.dropboxusercontent.com/cd/0/get/AIs3fSwXnnm-XwD3jSNT1zOkBTlEd9uxC2p1dibzGfnI1f2bOncX8sp5Wzm0rRLEKOE83hfCnRqSfPWbkpi4goRCUvRbbaaGtFCaEmrrrxdcvTRLbx0gdMbbxX6QyAFBi-u8LkN_MWvmMwzEfjn01ec5-2pNXe4A7KCHET24Snpev_qDGSqNsoSP_Ip-W8vkC9s/file [following]
--2018-06-12 17:59:53--  https://ucdb7c855b5e2daa601e6fc04f1d.dl.dropboxusercontent.com/cd/0/get/AIs3fSwXnnm-XwD3jSNT1zOkBTlEd9uxC2p1dibzGfnI1f2bOncX8sp5Wzm0rRLEKOE83hfCnRqSfPWbkpi4goRCUvRbbaaGtFCaEmrrrxdcvTRLbx0gdMbbxX6QyAFBi-u8LkN_MWvmMwzEfjn01ec5-2pNXe4A7KCHET24Snpev_qDGSqNsoSP_Ip-W8vkC9s/file
Resolving ucdb7c855b5e2daa601e6fc04f1d.dl.dropboxusercontent.com (ucdb7c855b5e2daa601e6fc04f1d.dl.dropboxuserconte

## 2. Database analysis
Now we will load our script for metadata loading & dataset preprocessing.

In [0]:
from preprocess import build_ds_meta

In [0]:
ds_meta = build_ds_meta()

for ds_meta_item in ds_meta:
    print('Quantity of %s: %s' % (ds_meta_item, ds_meta[ds_meta_item].count()[0]))

Quantity of bounding_boxes: 48562
Quantity of sizes: 48562
Quantity of hierarchy: 1010
Quantity of photographers: 48562
Quantity of image_class_labels: 48562
Quantity of classes: 1011
Quantity of images: 48562


The full database contains 48562 images distributed among 1011 classes in total. However the investigated subset consists of only 3010 bird pictures. They are representants of 50 species (classes). The quantity distribution of images in different classes is presented in the Chart 1. We see that it seems to be pretty uniform with 60 items on average.

![](http://mion.elka.pw.edu.pl/~mchmielo/tmp/wykres_hist.png)

Since different classes refer to appropriate bird species, which hierarchy is available as database metadata file, we were able to build classes dependency tree similar to the natural species hierarchy. The result graph, in SVG format, can be obtained under the following address [link](http://mion.elka.pw.edu.pl/~mchmielo/tmp/nbirds_hierarchy.svg).

Besides hierarchical class assignment database metadata contains some information about authors and dimensions of pictures. Also image bounding boxes are very important from analytical / computational point of view. It contain coordinates of minimal cropping areas which include whole animals. By applying cropping operation we are able to vastly reduce the amount of data send to the classifier (early discarding irrelevant pixels). Image file names are linked with the described metadata by unique *image_guid* properties. The relations between all described files are well presented in the Figure 2.

**Figure 2** Database metadata files ![Figure 2](https://raw.githubusercontent.com/zacateras/nn-nbirds/master/assets/dataset.png) 

## 3. Feature extraction with Gabor Filters + Perceptron

In order to prepare dataset images library for feature extraction we shall apply a few transformations. All pictures were cropped with bounding boxes defined in metadata, resized to 150 / 150 dimensions, converted to grayscale and transformed with Gabor filter. Hyperparameters of the filter were adjusted, so that different color and contrast bounderies were exposed. New images will be copied to the separate folder.

Example images before and after transformation are presented below:

<table>
    <tr>
        <td><img src="https://github.com/zacateras/nn-nbirds/blob/master/assets/sample_before.jpg?raw=1" alt="Sample Before" style="height: 300px;"/></td>
        <td><img src="https://github.com/zacateras/nn-nbirds/blob/master/assets/sample_after.jpg?raw=1" alt="Sample After" style="height: 300px;"/></td>
    </tr>
</table>

We set up stream image data generator to use prepared train folder and randomly apply a set of transforms - rotations and shifts on preprocessed images.

Then we initialized a sequence of NN layers. We fixed the number of layers to 2 hidden with Relu activation followed by a dense layer with 50 neurons (matching the number of classes) and sigmoid activation. As we discovered here, Keras is very flexible so that any rearrangements of the model are pretty fast and intuitive for a person with sufficient theoretical NN / DNN knowledge. Its API is also very readable.

Since our task was to test hidden layers of different sizes a set of models is created. We tested each combination of 64, 256, 1024 neuron quantities. Generated models were saved directly in a separate folder in the repository. Our best model yielded a bit more than 14% *categorical accuracy* and 39% of *top 5 categorical accuracy*. Detailed performances results are presented in the table below. The network architecture is encoded by appropriate names, following scheme: *d{number_of_neurons_in_the_layer_1}_{layer_1_activation}_d{number_of_neurons_in_the_layer_2}_{layer_2_activation}*.

<table>
    <img src="https://github.com/zacateras/nn-nbirds/blob/master/assets/perceptron_results.png?raw=1" alt="Perceptron Results" style="height: 400px;"/>
</table>

## 4. Inception v3 
For our baseline model, we have chosen Inception v3 model, described here: https://arxiv.org/abs/1512.00567.

![alt text](https://cloud.google.com/tpu/docs/images/inceptionv3onc--oview.png)

We used pretrained weights available in Keras framework and trained only top layers (visible in the 'Final part'  rectangle in the image above). This allowed us to save a lot of training time (there are about 24 milions of parameters in Inception v3 model and we've trained only two milions of them, which is about 8%).

In [0]:
width = height = 299

# First, we import model from keras framework

import keras

base_model = keras.applications.inception_v3.InceptionV3(include_top=False,
                                                         weights='imagenet',
                                                         input_shape=(width, height, 3))

# Then, we add top layers

from keras.layers import Dense, GlobalAveragePooling2D

x = base_model.output
x = GlobalAveragePooling2D()(x)
x = Dense(1024, activation='relu')(x)
predictions = Dense(50, activation='softmax')(x)

#We set pretrained layers as non trainable... :
for layer in base_model.layers:
  layer.trainable = False

model = keras.models.Model(inputs=base_model.input, outputs=predictions)
model.compile(optimizer='rmsprop', loss='categorical_crossentropy', metrics=['categorical_accuracy', 'top_k_categorical_accuracy'])


# And then we train and evaluate model
from preprocess import clip, resize
from generators import Generators


clip()
resize(width, height)

g = Generators(width=width, height=height, batch_size=32)

(g_train, cnt_train) = g.train()
(g_validation, cnt_validation) = g.validation()
(g_test, cnt_test) = g.test()

checkpointer = keras.callbacks.ModelCheckpoint(filepath='weights-{epoch:02d}.hdf5', verbose=1)

model.fit_generator(g_train, validation_data=g_validation, epochs=30, callbacks=[checkpointer])

models = sorted([m for m in os.listdir(os.getcwd()) if m.endswith("hdf5")])

def eval_model(model):
  m = keras.models.load_model(model)
  res = m.evaluate_generator(g_test, verbose=True)
  del m
  keras.backend.clear_session()

results = [eval_model(m) for m in models]
print(results)

We have managed to achieve 68% percent of categorical accuracy and 95% top_5_categorical_accuracy.

## 5. Convolutional Neural Network - custom models

Since in the second part of the project we had expected to do a lot more calculations, collect a lot more logs and test a lot more different neural network architectures, it was essential to prepare an efficient way of gather training outputs. We decided to prepare a script which automatically synchronizes a directory tree creating a copy of the results in Google Drive.

In [0]:
! pip install -U -q PyDrive
from gdrive import synchronize_all

Then we prepared script for executing the whole training process for one network parametrization. We assumed that the following parameters can be parametrized:

* maximum number of epochs (epochs),
* size of mini-batch (batch_size),
* width of input images (width),
* height of input images (height),
* number of neurons in descriptor layer (descriptor_size),
* dimensions of filters used in each convolution layer (filter_size),
* number of filters in each convolution layer (filter_number),
* regularization method used (kernel_reg - one from None, L1, L2).

Each training cycle involves a preprocessing phase: input immage clipping and resizing and a postprocessing phase: collecting logs and saved models.

In [0]:
import shutil
import os

from preprocess import clip, resize
from generators import Generators

from models.custom_dnn import custom_dnn

def process_custom(epochs, batch_size, width, height, descriptor_size, filter_size, filter_number, kernel_reg, kernel_reg_name):
    model_code = 'cnn_w%s_h%s_%s_descr_%s_x_%s_%s_filt_%s_batch_%s_reg_%s_epochs' % (width, height, descriptor_size, filter_number, filter_size[0], filter_size[1], batch_size, kernel_reg_name, epochs)
    model_path = 'cnn_models/%s.h5' % model_code
    model_logs_path = 'logs/%s' % model_code
    
    if os.path.exists(model_path):
        print('Model %s already exists! Skipping execution.' % model_code)
        return
    else:
        print('Training model %s...' % model_code)
    
    if os.path.exists(model_logs_path):
        shutil.rmtree(model_logs_path)
    os.makedirs(model_logs_path)
    os.makedirs('%s/tb' % model_logs_path)
    os.makedirs('%s/checkpoints' % model_logs_path) 
    
    # preprocess
    clip()
    resize(width, height)
    
    # generators
    g = Generators(width=width, height=height, batch_size=batch_size)

    (g_train, cnt_train) = g.train()
    (g_validation, cnt_validation) = g.validation()
    (g_test, cnt_test) = g.test()
    
    model = custom_dnn(
      width=width,
      height=height,
      output=cnt_train,
      descriptor_size=descriptor_size,
      filter_size=filter_size,
      filter_number=filter_number,
      kernel_reg=kernel_reg,
      kernel_reg_name=kernel_reg_name)

    cb_tb = keras.callbacks.TensorBoard(
      log_dir='%s/tb' % model_logs_path,
      histogram_freq=0,
      batch_size=batch_size,
      write_graph=True,
      write_grads=False,
      write_images=True)

    cb_mc = keras.callbacks.ModelCheckpoint(
      filepath='%s/checkpoints/weights-{epoch:02d}.hdf5' % model_logs_path,
      verbose=1)

    model.fit_generator(
      g_train,
      steps_per_epoch=2000 // batch_size,
      epochs=epochs,
      validation_data=g_validation,
      validation_steps=800 // batch_size,
      callbacks=[cb_tb]) # ModelCheckpointer was removed due to significant size of generated checkpoint files

    model.save(model_path)
    synchronize_all(drive)

Using TensorFlow backend.


In order to track changes in performance of the models on-line we exposed a TensorBoard service via localtunnel.

In [0]:
# Start tensorboard & expose via localtunnel
! npm install -g localtunnel > npm-lt.log
get_ipython().system_raw('tensorboard --logdir logs --host 0.0.0.0 --port 6006 &')
get_ipython().system_raw('lt --port 6006 > url.txt 2>&1 &')
! cat url.txt

your url is: https://fast-falcon-4.localtunnel.me

Then we started the whole training process. We prepared a list of different parametrizations taking a cross-product of three batch sizes *32, 64, 128*, three input image dimensions *64 x 64, 128 x 128, 256 x 256*, three descriptor sizes *256, 512, 1024*, two filter sizes *3 x 3, 5 x 5*, two filter numbers *16, 32*, three types of regularization *None, L1, L2*.

In [0]:
import keras
import os
from itertools import product

epochss=[150]
batch_sizes=[128, 32, 64]
dims=[(64, 64), (128, 128), (256, 256)]

descriptor_sizes=[512, 1024, 256]
filter_sizes=[(3, 3), (5, 5)]
filter_numbers=[32, 16]
keras_regs=[(None, None), (keras.regularizers.l1(0.01), 'L1(0.01)'), (keras.regularizers.l2(0.01), 'L2(0.01)')]

for item in product(epochss, batch_sizes, dims, descriptor_sizes, filter_sizes, filter_numbers, keras_regs):
    epochs = item[0]
    batch_size = item[1]
    width = item[2][0]
    height = item[2][1]
    
    descriptor_size = item[3]
    filter_size = item[4]
    filter_number = item[5]
    kernel_reg = item[6][0]
    kernel_reg_name = item[6][1]
    
    process_custom(epochs, batch_size, width, height, descriptor_size, filter_size, filter_number, kernel_reg, kernel_reg_name)

Training model cnn_w128_h128_512_descr_32_x_3_3_filt_128_batch_L2(0.01)_reg_150_epochs...


Trained models were also immediately commited to the repository.

In [0]:
! git config --global user.email "USERMAIL"
! git config --global user.name "USERNAME"
! git add cnn_models/*
! git commit -m "cnn_models"
! git remote set-url origin https://USERNAME:PASSWORD@github.com/zacateras/nn-nbirds.git
! git push

[master 921e27a] cnn_models
 1 file changed, 0 insertions(+), 0 deletions(-)
 create mode 100644 cnn_models/cnn_512_descr_32_x_3_3_filt_128_batch_None_reg_120_epochs.h5
Counting objects: 4, done.
Delta compression using up to 2 threads.
Compressing objects: 100% (4/4), done.
Writing objects: 100% (4/4), 13.20 MiB | 5.81 MiB/s, done.
Total 4 (delta 1), reused 0 (delta 0)
remote: Resolving deltas: 100% (1/1), completed with 1 local object.
To https://github.com/zacateras/nn-nbirds.git
   c63d25a..921e27a  master -> master


Then we loaded all obtained models and tested them against the training set prepared at the beginning of the project. The results were stored in *cnn_models / results.csv* file.

In [0]:
from keras.models import load_model
from generators import Generators
import pandas as pd
import os
import re

records = None

for filename in os.listdir('cnn_models'):
    if not filename.endswith('.h5'):
        continue
    
    if records is not None and 'filename' in records.columns:
        if filename in [item for item in records['filename']]:
            print('Skipping for %s...' % filename)
            continue
        
    print('Processing %s...' % filename)

    regex = re.compile('cnn_w(\d+)_h(\d+)_(\d+)_descr_(\d+)_x_(\d+)_(\d+)_filt_(\d+)_batch_(L1\(0.01\)|L1\(0.0001\)|L2\(0.01\)|L2\(0.0001\)|None)_reg_(\d+)_epochs')
    match = regex.match(filename)
    record = {
        "filename": filename,
        "width": int(match.group(1)),
        "height": int(match.group(2)),
        "descriptor_size": int(match.group(3)),
        "filter_count": int(match.group(4)),
        "filter_width": int(match.group(5)),
        "filter_height": int(match.group(6)),
        "mini_batch": int(match.group(7)),
        "regularization": match.group(8),
        "epochs": int(match.group(9))
    }

    [record['loss'], record['categorical_accuracy'], record['top5_categorical_accuracy']] = load_model('cnn_models/%s' % filename) \
        .evaluate_generator(Generators(record['width'], record['height'], record['mini_batch']).test()[0])

    if records is None:
        records = pd.DataFrame(record, index=[0])
    else:
        records = records.append(pd.DataFrame(record, index=[0]))

df = pd.DataFrame(records).sort_values('categorical_accuracy', ascending=False)
df.to_csv('cnn_models/results.csv')

## 6. Convolutional Neural Network - custom results

In the table below we present top performing models according to the tests we performed. Due to unavailability of Collaboratory in the last phase of our project we were unable to execute all the tests that we scheduled before. In consequence we tested ~75 out of 300 configurations in total. This is also an important notice for our future use. The Collaboratory service is not intended to be used as a background task execution platform. Intensive continous usage can trigger resource limiting policies, significantly reduce the performance of training.

In [0]:
df[df['categorical_accuracy'] > 0.05]\
    [['epochs', 'width', 'height', 'descriptor_size', 'mini_batch', 'filter_count', 'filter_height', 'filter_width', 'regularization', 'categorical_accuracy', 'top5_categorical_accuracy']]\
    .sort_values(['categorical_accuracy'], ascending=[0])\
    .head(8)

epochs  width  height  descriptor_size  mini_batch  filter_count  \
0     150    128     128              512         128            32   
0     250     64      64              512         128            32   
0     150     64      64              512         128            32   
0     150     64      64              512         128            32   
0     150     64      64             1024          32            32   
0     150     64      64              256         128            32   
0     120     64      64              512         128            32   
0     150     64      64             1024         128            32   

   filter_height  filter_width regularization  categorical_accuracy  \
0              3             3           None              0.617792   
0              3             3           None              0.589786   
0              5             5           None              0.574959   
0              3             3           None              0.570016   
0              3             3           None              0.566722   
0              3             3           None              0.561779   
0              3             3           None              0.556837   
0              3             3           None              0.551895   

   top5_categorical_accuracy  
0                   0.853377  
0                   0.841845  
0                   0.833608  
0                   0.817133  
0                   0.840198  
0                   0.812191  
0                   0.838550  
0                   0.817133

The best model we were able to obtain is the one shown in the picture below. It was trained with images rescaled to 128px x 128px:

![alt text](https://raw.githubusercontent.com/zacateras/nn-nbirds/master/images/my_cnn.png)

In this case we achieved categorical accuracy at the level of **62%** on the test set. The ROC curve (averaged) of the model is presented in the figure below, on the left:

<table>
    <tr>
        <td><img src="https://github.com/zacateras/nn-nbirds/blob/master/images/index.png?raw=1" alt="" style="height: 300px;"/></td>
        <td><img src="https://github.com/zacateras/nn-nbirds/blob/master/images/index10.png?raw=1" alt="" style="height: 300px;"/></td>
    </tr>
</table>

Since training of 128 x 128 models took long, we performed the rest of the experiments for nets operating on 64 x 64 input images. The topology the network was not changed. For such images we observed the highest categorical accuracy at the level of **59%** on the test set, when the training was continued until 250s' epoch and **57%** for natural 150 epochs experiments. We assumed this model as the baseline for the next few analytical experiments. The ROC curve of this model is the one above on the right.

Changing the size of filters from (3 x 3) to (5 x 5) resulted in increasing the accuracy by **0.5%** up to **57.5%**. Different number of the filters used (16, 16, 32, 32, 32, 32) yielded in worse results - **53%**. The experiment of taking different mini-batch sizes resulted in the conclusion - as we lowered this parameter to 64 and then to 32, the categorical accuracy dropped dramatically. For 32 value used, the networks were mostly unable to fit to data - **~2%** accuracy. We also examined performance of 256, 1024. In comparison to 512 descriptor we observed a slight decrease of performance indicators - **55.2%** for 1024 and **54.9%** for 256.

In the last experiment we checked if using L1 and L2 regularizations can prevent our model with overfitting. Surprissingly, applying both regularisers - as they are given in Keras out of the box - with lambda set to either 0.01 or 0.0001 resulted in absolute decrease of model accuracy - **2%**. ROC curves comparison is presented below:

![alt text](https://raw.githubusercontent.com/zacateras/nn-nbirds/master/images/index78a.png)


## 7.  Support Vector Machines
The third part of second phase of our project was to train Support Vector Machines on descriptors obtained from the selected CNN model. We have used the model - *cnn_w128_h128_512_descr_32_x_3_3_filt_128_batch_None_reg_150_epochs*, with baseline accuracy of 61%.

SVMs were implemented using Scikit Learn framework, with exception to exponential kernel function. Since it is not as popular as similar Gaussian function, it is not implemented in Sckit Learn yet. Instead we used implementation found here: https://github.com/gmum/pykernels (code from this repository was pulled to our solution and then just slightly adjusted).

Using two different SVMs - with polynomial and exponential kernel we have manged to achieve accordingly accuracy of 58% and 64%. We measured an impact of the penalty parameter (C) on classification results. Our measurements are depicted below:

For exponential kernel the best results was observed using sigma parameter equal to 7.

<table>
    <tr>
        <td><img src="./assets/plt_exp.png" alt="" style="height: 300px;"/></td>
        <td><img src="./assets/plt_poly.png" alt="" style="height: 300px;"/></td>
    </tr>
</table>

The penalty parameter balances tradeoff between number of missclassified samples versus simplicity of decision surface. Low C parameter favours simpler boundaries, and high allows to fit them tightly to training data. Intuitivly, too high values of C can result in overfitting. However, for our results, overfitting cannot be observed, what may mean that the data obtained from CNN descriptors are easily linearly separable in transformed feature space.

## 8. Tools evaluation

### Collab

Google Collab is very useful for creating and testing small neural network achitectures. It provides GPU for free, which allowed us to train new models in time and cont efficient manner. On the other hand, it has some serious drawbacks, which are unacceptable in professional work. The key shortcomings of Collab are:

- instability: getting a decent runtime with GPU sometimes needs a few restarts, sometimes it boots with python 3 instead of 2, or using GPU with only 256 MB of internal memory,
- problems with data synchronization: Collab is cloud based and its storage isn't prsistent, and getting data out of it requires some work, i.e. setting up synchronization with Google drive, instead of just copying files with using SCP

### Keras

Using Keras for network implementation is a pleasure. Popular neural network achitectures with pretrained weights are available out of the box. Thanks to Tensorflow backend training is fast, in comparison to market competitors. The only problem with Keras is relative immaturity. Many branches of the software are insufficiently documented.

## 9. Conclusions

For image classification problem, Convolutional Neural Networks are far more effective than perceptrons. Using pretrained state of the art model one can achive satysfying results with very short network training. In some cases, CNN results can be further boosted by treating them as descriptor generators and using Support Vector Machines to classify descriptors.

## References

#### Google Colab
* [Colab Teaser](https://hackernoon.com/train-your-machine-learning-models-on-googles-gpus-for-free-forever-a41bd309d6ad)
* [Simple Colab Example](https://towardsdatascience.com/fast-ai-lesson-1-on-google-colab-free-gpu-d2af89f53604)
* [Extensive Colab Example](https://medium.com/deep-learning-turkey/google-colab-free-gpu-tutorial-e113627b9f5d)

#### Gabor Filters
* [Practical overview of Gabor Filters](https://cvtuts.wordpress.com/2014/04/27/gabor-filters-a-practical-overview/)
* [Library for testing various filter configurations](http://nbviewer.jupyter.org/github/bicv/LogGabor/blob/master/LogGabor.ipynb)
* [Example of applying OpenCV implementation in Python](https://gist.github.com/kendricktan/93f0da88d0b25087d751ed2244cf770c)
* [Feature extraction in face recognition task](https://pdfs.semanticscholar.org/3cf9/e4bc906ee4d15b69b34db413f8e319692e3b.pdf)
* [Feature extraction in car recognition task](http://scholar.google.pl/scholar_url?url=http://www.academia.edu/download/3448045/Car_recognition_using_gabor_filter_feature_extraction.doc&hl=pl&sa=X&scisig=AAGBfm2C3mPP89cH7bXDsFiXMDNPLIY27Q&nossl=1&oi=scholarr&ved=0ahUKEwizv5bdgdTaAhXEGuwKHXy-AY0QgAMIKSgCMAA)

#### DNNs in Keras
* [Keras documentation](https://keras.io/)
* [Keras example](https://blog.keras.io/building-powerful-image-classification-models-using-very-little-data.html)